# Бинарная классификация мошеннических транзакций
# Предварительный анализ данных

In [2]:
import pandas as pd

TRAIN_PATH = "train.csv"
TEST_PATH = "test.csv"

df = pd.read_csv(TRAIN_PATH)

print("Размер таблицы:", df.shape)  # (строки, столбцы)
display(df.head())  # первые 5 строк


Размер таблицы: (219129, 32)


id  Time        V1        V2        V3        V4        V5        V6  \
0   0   0.0  2.074329 -0.129425 -1.137418  0.412846 -0.192638 -1.210144   
1   1   0.0  1.998827 -1.250891 -0.520969 -0.894539 -1.122528 -0.270866   
2   2   0.0  0.091535  1.004517 -0.223445 -0.435249  0.667548 -0.988351   
3   3   0.0  1.979649 -0.184949 -1.064206  0.120125 -0.215238 -0.648829   
4   4   0.0  1.025898 -0.171827  1.203717  1.243900 -0.636572  1.099074   

         V7        V8  ...       V21       V22       V23       V24       V25  \
0  0.110697 -0.263477  ... -0.334701 -0.887840  0.336701 -0.110835 -0.291459   
1 -1.029289  0.050198  ...  0.054848 -0.038367  0.133518 -0.461928 -0.465491   
2  0.948146 -0.084789  ... -0.326725 -0.803736  0.154495  0.951233 -0.506919   
3 -0.087826 -0.035367  ... -0.095514 -0.079792  0.167701 -0.042939  0.000799   
4 -0.938651  0.569239  ...  0.099157  0.608908  0.027901 -0.262813  0.257834   

        V26       V27       V28  Amount  Class  
0  0.207733 -0.076576 -0.059577    1.98      0  
1 -0.464655 -0.009413 -0.038238   84.00      0  
2  0.085046  0.224458  0.087356    2.69      0  
3 -0.096148 -0.057780 -0.073839    1.00      0  
4 -0.252829  0.108338  0.021051    1.00      0  

[5 rows x 32 columns]

In [3]:
df.info(memory_usage="deep")



<class 'pandas.core.frame.DataFrame'>
RangeIndex: 219129 entries, 0 to 219128
Data columns (total 32 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   id      219129 non-null  int64  
 1   Time    219129 non-null  float64
 2   V1      219129 non-null  float64
 3   V2      219129 non-null  float64
 4   V3      219129 non-null  float64
 5   V4      219129 non-null  float64
 6   V5      219129 non-null  float64
 7   V6      219129 non-null  float64
 8   V7      219129 non-null  float64
 9   V8      219129 non-null  float64
 10  V9      219129 non-null  float64
 11  V10     219129 non-null  float64
 12  V11     219129 non-null  float64
 13  V12     219129 non-null  float64
 14  V13     219129 non-null  float64
 15  V14     219129 non-null  float64
 16  V15     219129 non-null  float64
 17  V16     219129 non-null  float64
 18  V17     219129 non-null  float64
 19  V18     219129 non-null  float64
 20  V19     219129 non-null  float64
 21  V20     21

In [4]:
# проверка дубликатов строк
n_dupes = df.duplicated().sum()
print(f"Дубликаты: {n_dupes} ({n_dupes / len(df):.4%})")


Дубликаты: 0 (0.0000%)


In [5]:
# пропуски по столбцам
na_counts = df.isna().sum()
na_perc = na_counts / len(df) * 100
summary_na = (
    pd.DataFrame({"na_count": na_counts, "na_%": na_perc})
    .query("na_count > 0")
    .sort_values("na_count", ascending=False)
)

if summary_na.empty:
    print("✔️ Пропуски не обнаружены.")
else:
    display(summary_na)


✔️ Пропуски не обнаружены.


In [6]:
# распределение целевого признака
target_counts = df["Class"].value_counts().sort_index()
target_perc = target_counts / len(df) * 100
print("Распределение Class (0 — legit, 1 — fraud):")
display(pd.concat([target_counts, target_perc.rename("%")], axis=1))


Распределение Class (0 — legit, 1 — fraud):


count          %
Class                   
0      218660  99.785971
1         469   0.214029

In [7]:
# базовые статистики по числовым полям
numeric_stats = df.describe().T  # .T = транспонировать для удобства
display(numeric_stats)


count           mean           std        min           25%  \
id      219129.0  109564.000000  63257.237906   0.000000  54782.000000   
Time    219129.0   62377.415376  25620.348569   0.000000  47933.000000   
V1      219129.0       0.096008      1.395425 -29.807725     -0.846135   
V2      219129.0       0.048345      1.159805 -44.247914     -0.573728   
V3      219129.0       0.592102      1.132884 -19.722872     -0.027154   
V4      219129.0       0.069273      1.253125  -5.263650     -0.769256   
V5      219129.0      -0.161555      1.069530 -37.591259     -0.847346   
V6      219129.0       0.133688      1.202411 -25.659750     -0.631835   
V7      219129.0      -0.128224      0.817207 -31.179799     -0.646730   
V8      219129.0       0.149534      0.716212 -28.903442     -0.095948   
V9      219129.0      -0.048337      1.054143  -8.756951     -0.711444   
V10     219129.0      -0.039758      0.821889 -22.092656     -0.499563   
V11     219129.0       0.153632      0.976946  -4.190145     -0.576969   
V12     219129.0      -0.061038      0.998470 -16.180165     -0.476890   
V13     219129.0       0.014330      1.039145  -4.373778     -0.671601   
V14     219129.0       0.067649      0.801335 -15.585021     -0.329905   
V15     219129.0       0.108643      0.891613  -4.155728     -0.461596   
V16     219129.0       0.013650      0.786654 -11.778839     -0.461077   
V17     219129.0       0.036815      0.691709 -20.756768     -0.406675   
V18     219129.0      -0.033927      0.784454  -7.456060     -0.496990   
V19     219129.0      -0.008302      0.739928  -4.281628     -0.463035   
V20     219129.0       0.009708      0.439521 -18.679066     -0.167927   
V21     219129.0      -0.031064      0.422777 -14.689621     -0.190418   
V22     219129.0      -0.050852      0.597812  -8.748979     -0.473099   
V23     219129.0      -0.050531      0.318175 -11.958588     -0.174478   
V24     219129.0      -0.002992      0.593100  -2.836285     -0.332540   
V25     219129.0       0.124005      0.406741  -3.958591     -0.126080   
V26     219129.0       0.009881      0.473867  -1.858672     -0.318330   
V27     219129.0       0.014034      0.233355  -9.234767     -0.050983   
V28     219129.0       0.017313      0.164859  -4.551680     -0.009512   
Amount  219129.0      66.359803    150.795017   0.000000      5.990000   
Class   219129.0       0.002140      0.046214   0.000000      0.000000   

                  50%            75%            max  
id      109564.000000  164346.000000  219128.000000  
Time     63189.000000   77519.000000  120580.000000  
V1           0.385913       1.190661       2.430494  
V2           0.046937       0.814145      16.068473  
V3           0.735895       1.306110       6.145578  
V4           0.064856       0.919353      12.547997  
V5          -0.229929       0.356856      34.581260  
V6          -0.087778       0.482388      16.233967  
V7          -0.098970       0.385567      39.824099  
V8           0.111219       0.390976      18.270586  
V9          -0.131323       0.583715      13.423914  
V10         -0.106034       0.403967      15.878405  
V11          0.090545       0.917392       9.417789  
V12          0.087649       0.608480       5.406614  
V13         -0.016837       0.695547       5.976265  
V14          0.049266       0.460837       6.078453  
V15          0.178975       0.791255       4.693323  
V16          0.054550       0.531777       5.834992  
V17         -0.013949       0.410978       8.845303  
V18         -0.039451       0.446448       4.847887  
V19         -0.002935       0.455718       4.090974  
V20         -0.037702       0.126750      15.407839  
V21         -0.042858       0.109187      22.062945  
V22         -0.032856       0.354910       6.163541  
V23         -0.063307       0.060221      12.734391  
V24          0.038708       0.394566       4.572739  
V25          0.145934       0.402926       3.111624  
V26         -0.086388       0.253869       3.402344  
V2

In [8]:
# быстрый вывод ключевых фактов одной строкой (опционально)
print(f"""
👉 Строк: {len(df):,}
👉 Столбцов: {df.shape[1]}
👉 Пропуски: {df.isna().any().sum()} столбцов с NaN
👉 Дубликаты: {n_dupes}
👉 Fraud rate: {target_perc.loc[1]:.3f}%  (≈ 1 к {int(target_counts.loc[0] / target_counts.loc[1]):.0f})
""")



👉 Строк: 219,129
👉 Столбцов: 32
👉 Пропуски: 0 столбцов с NaN
👉 Дубликаты: 0
👉 Fraud rate: 0.214%  (≈ 1 к 466)



# Выводы

Пропусков нет → этап обработки пропусков упрощается: достаточно подтвердить это после трансформаций.

Все столбцы числовые. Категориальных признаков нет, поэтому кодирование (One-Hot, Ordinal и т. п.) не требуется.

Сильный дисбаланс классов (~0.2 %) — при моделировании понадобится:

- стратификация при сплите,

- метрики, устойчивые к дисбалансу (AUC-ROC, PR-AUC),

- техники балансировки (взвешивание классов, oversampling/undersampling) — опционально.

- id следует исключить из обучающих признаков (это индекс).

- Нормализация/стандартизация желательна: диапазоны Amount и PCA-компонент различаются; выберем метод (например, StandardScaler) на этапе подготовки данных.



# Пропуски в данных

In [9]:
df_train = pd.read_csv(TRAIN_PATH)
df_test = pd.read_csv(TEST_PATH)

print("✅ train shape:", df_train.shape)


✅ train shape: (219129, 32)


In [10]:
# функция проверки пропусков
def missing_report(df, name="dataframe"):
    """Возвращает таблицу пропусков (count, %) и печатает краткое резюме."""
    na_count = df.isna().sum()
    na_pct = na_count / len(df) * 100
    report = (
        pd.DataFrame({"na_count": na_count, "na_%": na_pct})
        .query("na_count > 0")
        .sort_values("na_count", ascending=False)
    )
    if report.empty:
        print(f"✔️  В '{name}' пропусков не обнаружено.")
    else:
        print(f"⚠️  Пропуски в '{name}':")
        display(report)
    return report


In [11]:
# отчёты по пропускам
missing_train = missing_report(df_train, "train")


✔️  В 'train' пропусков не обнаружено.


In [12]:
print(f"""
Итого:
• Столбцов с NaN в train: {len(missing_train)}
""")



Итого:
• Столбцов с NaN в train: 0



# Т.к. пропусков нет, можно продолжать работу с данными.

In [13]:
df_test = pd.read_csv(TEST_PATH)
print("✅ test.csv shape:", df_test.shape)
display(df_test.head())


✅ test.csv shape: (146087, 31)


id      Time        V1        V2        V3        V4        V5  \
0  219129  120580.0  2.115519 -0.691809 -1.305514 -0.685655 -0.641265   
1  219130  120580.0  1.743525 -1.681429 -0.547387 -1.061113 -0.695825   
2  219131  120580.0  2.205568 -1.571445 -0.238965 -1.662517 -1.652324   
3  219132  120580.0  1.989728 -0.972909 -1.938259 -1.440129 -0.166855   
4  219133  120580.0 -1.943548 -1.668761  0.363601 -0.977610  2.684779   

         V6        V7        V8  ...       V20       V21       V22       V23  \
0 -0.764784 -0.924262 -0.023030  ...  0.067367  0.241708  0.682524  0.037769   
1  2.458824 -1.632859  1.073529  ...  0.441788  0.543278  1.294571  0.309541   
2 -0.054701 -1.682064  0.105613  ... -0.366906 -0.131527  0.086623  0.291375   
3 -0.794048  0.252889 -0.399789  ... -0.049136 -0.080115 -0.010732 -0.038550   
4 -2.037681  0.039709 -0.048895  ...  0.391627  0.083389 -0.306918  0.247822   

        V24       V25       V26       V27       V28  Amount  
0 -0.546859 -0.123055 -0.084889  0.004720 -0.021944   29.95  
1  3.703925 -0.242579  0.068708  0.002629  0.064690  163.50  
2  0.739087 -0.543006 -0.297813  0.043699 -0.037855   16.00  
3  0.656830  0.343470 -0.627529 -0.024338 -0.036143  120.98  
4 -0.391799 -0.790716 -0.025706  0.330758  0.335537    1.98  

[5 rows x 31 columns]

In [14]:
#  функции для определения категориальных признаков
def find_categorical_columns(
        df: pd.DataFrame,
        max_unique_int: int = 20,
        max_unique_frac: float = 0.05,
) -> list[str]:
    """
    Возвращает список столбцов, которые можно считать категориальными.

    • object / category dtype — всегда.
    • int/float dtype — если количество уникальных
      1) ≤ max_unique_int И
      2) доля уникальных ≤ max_unique_frac (по умолчанию ≤5 %).
    """
    cat_cols = []

    for col in df.columns:
        dtype = df[col].dtype

        # 1. Явно категориальные типы
        if dtype == "object" or str(dtype).startswith("category"):
            cat_cols.append(col)
            continue

        # 2. Числовые с ограниченным числом уникальных значений
        if pd.api.types.is_numeric_dtype(dtype):
            uniq = df[col].nunique(dropna=False)
            uniq_frac = uniq / len(df)
            if (uniq <= max_unique_int) and (uniq_frac <= max_unique_frac):
                cat_cols.append(col)

    return cat_cols


In [15]:
# поиск и отчёт
categorical_cols = find_categorical_columns(df_test)

if categorical_cols:
    print("🗂️ Найдены потенциально категориальные признаки:")
    for c in categorical_cols:
        uniq = df_test[c].nunique(dropna=False)
        print(f"• {c:<15}  (уникальных: {uniq})")
else:
    print("✔️  Категориальные признаки не обнаружены.")


✔️  Категориальные признаки не обнаружены.


In [16]:
# сводка по уникальным значениям
# Показывает топ-10 признаков с наименьшим числом уникальных значений
(
    df_test.apply(lambda s: s.nunique(dropna=False))
    .sort_values()
    .head(10)
    .to_frame("unique_values")
    .style.format(thousands=" ")
)


# Нормирование данных
Метод нормирования: StandardScaler (Z-score: (x − μ)/σ )

- Центрирует и приводит всё к дисперсии 1.
- Устойчив для признаков, уже похожих на нормальное распределение (PCA-компоненты V1–V28).


In [17]:
#  выделяем признаки и целевую переменную
from sklearn.preprocessing import StandardScaler
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
import joblib  # чтобы сохранить scaler на будущее

TARGET_COL = "Class"
ID_COL = "id"

X_train = df_train.drop(columns=[TARGET_COL, ID_COL])
y_train = df_train[TARGET_COL]

X_test = df_test.drop(columns=[ID_COL])  # в test нет Class




In [18]:
# создаём пайплайн со StandardScaler
num_features = X_train.columns.tolist()  # все оставшиеся числовые

preprocess = ColumnTransformer(
    transformers=[
        ("num", StandardScaler(), num_features)
    ],
    remainder="drop"
)

pipeline = Pipeline([
    ("scaler", preprocess)
])


In [19]:
# 🟢 Ячейка 5 — фит и трансформ
X_train_scaled = pipeline.fit_transform(X_train)
X_test_scaled = pipeline.transform(X_test)

print("Масштабированные формы:", X_train_scaled.shape, X_test_scaled.shape)


Масштабированные формы: (219129, 30) (146087, 30)


In [20]:
# преобразуем обратно в DataFrame
scaled_cols = num_features
df_train_scaled = pd.DataFrame(X_train_scaled, columns=scaled_cols)
df_train_scaled[TARGET_COL] = y_train.values  # возвращаем цель

df_test_scaled = pd.DataFrame(X_test_scaled, columns=scaled_cols)
df_train_scaled.head()


Time        V1        V2        V3        V4        V5        V6  \
0 -2.434688  1.417723 -0.153277 -1.526656  0.274174 -0.029063 -1.117617   
1 -2.434688  1.363615 -1.120222 -0.982513 -0.769129 -0.898502 -0.336453   
2 -2.434688 -0.003205  0.824427 -0.719888 -0.402612  0.775205 -0.933160   
3 -2.434688  1.349872 -0.201150 -1.462031  0.040580 -0.050193 -0.650792   
4 -2.434688  0.666386 -0.189836  0.539876  0.937360 -0.444137  0.802877   

         V7        V8        V9  ...       V21       V22       V23       V24  \
0  0.292364 -0.576662  0.749882  ... -0.718199 -1.400090  1.217046 -0.181829   
1 -1.102617 -0.138696 -0.058446  ...  0.203208  0.020885  0.578455 -0.773794   
2  1.317136 -0.327171  0.005986  ... -0.699333 -1.259402  0.644384  1.608883   
3  0.049435 -0.258165  0.886196  ... -0.152446 -0.048410  0.685889 -0.067352   
4 -0.991706  0.586008  0.702944  ...  0.308014  1.103629  0.246506 -0.438073   

        V25       V26       V27       V28    Amount  Class  
0 -1.021451  0.417528 -0.388293 -0.466403 -0.426937      0  
1 -1.449321 -1.001415 -0.100477 -0.336961  0.116982      0  
2 -1.551175  0.158620  0.901735  0.424868 -0.422228      0  
3 -0.302912 -0.223754 -0.307748 -0.552913 -0.433436      0  
4  0.329029 -0.554398  0.404124  0.022671 -0.433436      0  

[5 rows x 31 columns]

In [21]:
# сохраняем scaler и/или готовые таблицы
joblib.dump(pipeline, "standard_scaler_pipeline.joblib")
df_train_scaled.to_csv("train_scaled.csv", index=False)
df_test_scaled.to_csv("test_scaled.csv", index=False)


# Выбросы

In [22]:
from scipy import stats
import numpy as np


num_cols = df.select_dtypes(include=[np.number]).columns.tolist()



In [23]:
# расчёт z-score и маски выбросов
z_scores = np.abs(stats.zscore(df[num_cols], axis=0, nan_policy="omit"))
outlier_mask = z_scores > 3  # True, где |z| > 3


In [24]:
# сводная таблица по выбросам
summary = (
    pd.DataFrame({
        "feature": num_cols,
        "outliers": outlier_mask.sum(axis=0),
    })
    .assign(
        outlier_pct=lambda d: d["outliers"] / len(df) * 100,
        min_val=lambda d: [df[col].min() for col in d["feature"]],
        max_val=lambda d: [df[col].max() for col in d["feature"]],
        method="3-σ"
    )
    .loc[lambda d: d["outliers"] > 0]  # оставляем только признаки с выбросами
    .sort_values("outliers", ascending=False)
    .reset_index(drop=True)
)

display(summary)  # красивый вывод в Jupyter


feature  outliers  outlier_pct    min_val      max_val method
0       V6      4347     1.983763 -25.659750    16.233967    3-σ
1   Amount      3695     1.686221   0.000000  7475.000000    3-σ
2      V27      3526     1.609098  -9.234767    13.123618    3-σ
3      V20      3404     1.553423 -18.679066    15.407839    3-σ
4      V23      3061     1.396894 -11.958588    12.734391    3-σ
5       V8      2907     1.326616 -28.903442    18.270586    3-σ
6      V28      2785     1.270941  -4.551680    23.263746    3-σ
7       V2      2686     1.225762 -44.247914    16.068473    3-σ
8      V21      2523     1.151377 -14.689621    22.062945    3-σ
9      V14      2407     1.098440 -15.585021     6.078453    3-σ
10     V12      2087     0.952407 -16.180165     5.406614    3-σ
11     V17      2068     0.943736 -20.756768     8.845303    3-σ
12      V1      2062     0.940998 -29.807725     2.430494    3-σ
13      V5      1760     0.803180 -37.591259    34.581260    3-σ
14      V7      1760     0.803180 -31.179799    39.824099    3-σ
15      V3      1664     0.759370 -19.722872     6.145578    3-σ
16     V10      1573     0.717842 -22.092656    15.878405    3-σ
17     V19      1445     0.659429  -4.281628     4.090974    3-σ
18     V25      1370     0.625203  -3.958591     3.111624    3-σ
19      V4      1364     0.622464  -5.263650    12.547997    3-σ
20     V16      1179     0.538039 -11.778839     5.834992    3-σ
21     V18      1150     0.524805  -7.456060     4.847887    3-σ
22      V9      1013     0.462285  -8.756951    13.423914    3-σ
23     V22       887     0.404784  -8.748979     6.163541    3-σ
24     V15       825     0.376491  -4.155728     4.693323    3-σ
25     V13       789     0.360062  -4.373778     5.976265    3-σ
26     V26       626     0.285676  -1.858672     3.402344    3-σ
27     V24       566     0.258295  -2.836285     4.572739    3-σ
28   Class       469     0.214029   0.000000     1.000000    3-σ
29     V11       226     0.103136  -4.190145     9.417789    3-σ

Всего 16 из 30 числовых признаков имеют ≥ 1 % наблюдений за пределами 3 стандартных отклонений; остальные находятся в диапазоне 0.3–0.9 %.

Сводно:

29 270 транзакций (13.36 % выборки) содержат хотя бы один «сигма-выброс».

Наиболее «тяжёлый хвост» — у PCA-компонент V6, V27, V20, V23.

Amount выражено в евро; несмотря на долгий хвост (до €7 475), по 3-σ-критерию выбросами считается лишь 1.7 %.

Однако по правилу IQR*1.5 все суммы > €163 (или < 0 €) дают 20 617 выбросов (9.4 %).

# Варианты нормализации

| Вариант                                    | Преимущества                                                                                                   | Недостатки                                                              |
| ------------------------------------------ | -------------------------------------------------------------------------------------------------------------- | ----------------------------------------------------------------------- |
| **RobustScaler** (медиана / IQR)           | Снижает влияние длинных хвостов `Amount`, `V6` и др.                                                           | Лёгкое искажение признаков, где распределение уже близко к нормальному. |
| **Клиппинг** (Winsorize ±3 σ или IQR\*1.5) | Убирает экстремальные значения без удаления строк.                                                             | Потеря информации; порог выбирается эвристически.                       |
| **Оставить как есть**                      | PCA-признаки часто ведут себя как норм. распределение с редкими «краями»; модели на основе деревьев устойчивы. | Линейные модели чувствительны; может ухудшить масштабирование.          |



In [39]:
# Для тренировочных данных
# Взаимодействия и полиномы
df_train_scaled['V6_V27'] = df_train_scaled['V6'] * df_train_scaled['V27']
df_train_scaled['Amount_sq'] = df_train_scaled['Amount'] ** 2

# Бинарные маркеры выбросов (3σ)
for col in ['V6', 'Amount']:
    df_train_scaled[f'IsOutlier_{col}'] = (np.abs(df_train_scaled[col]) > 3).astype(int)

# Логарифмирование суммы
df_train_scaled['Log_Amount'] = np.log1p(df_train_scaled['Amount'])

# Кластеризация K-means
kmeans = KMeans(n_clusters=5, random_state=42)
kmeans.fit(df_train_scaled[['V6', 'V20', 'V23', 'V27']])
df_train_scaled['Cluster'] = kmeans.predict(df_train_scaled[['V6', 'V20', 'V23', 'V27']])

# Для тестовых данных

df_test_scaled['V6_V27'] = df_test_scaled['V6'] * df_test_scaled['V27']
df_test_scaled['Amount_sq'] = df_test_scaled['Amount'] ** 2

# Бинарные маркеры выбросов
for col in ['V6', 'Amount']:  # Исправлено: двоеточие
    df_test_scaled[f'IsOutlier_{col}'] = (np.abs(df_test_scaled[col]) > 3).astype(int)

# Логарифмирование суммы
df_test_scaled['Log_Amount'] = np.log1p(df_test_scaled['Amount'])

# Кластеризация (используем обученный кластеризатор!)
df_test_scaled['Cluster'] = kmeans.predict(df_test_scaled[['V6', 'V20', 'V23', 'V27']])

Взаимодействия (V6_V27), полиномы (Amount_sq), бинарные маркеры выбросов (IsOutlier_V6), логарифмированная сумма (Log_Amount) и кластеры (Cluster) — созданы для выявления нелинейных зависимостей и аномалий.

In [40]:
# Блок 2: Отбор признаков
from sklearn.feature_selection import SelectFromModel
from lightgbm import LGBMClassifier

# Метод важности признаков
X = df_train_scaled.drop(columns=['Class'])
y = df_train_scaled['Class']

In [41]:
# LightGBM Feature Importance
lgb = LGBMClassifier(random_state=42, class_weight='balanced')
lgb.fit(X, y)
importance = pd.Series(lgb.feature_importances_, index=X.columns)
selected_features_imp = importance.nlargest(15).index.tolist()

[LightGBM] [Info] Number of positive: 469, number of negative: 218660
[LightGBM] [Info] Auto-choosing col-wise multi-threading, the overhead of testing was 0.108421 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 8424
[LightGBM] [Info] Number of data points in the train set: 219129, number of used features: 36
[LightGBM] [Info] [binary:BoostFromScore]: pavg=0.500000 -> initscore=-0.000000
[LightGBM] [Info] Start training from score -0.000000


In [28]:
from sklearn.linear_model import LogisticRegression
from sklearn.feature_selection import SelectFromModel
from sklearn.datasets import make_classification

In [51]:
# L1-регуляризация
selector = SelectFromModel(
    estimator=LogisticRegression(
        penalty='l1',
        solver='saga',
        class_weight='balanced',
        max_iter=5000,
        random_state=42
    ),
    threshold="median"
)



In [52]:
selector.fit(X, y)
selected_features_l1 = X.columns[selector.get_support()]

/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_sag.py:348: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(


In [53]:
print("Выбрано признаков:", sum(selector.get_support()))

Выбрано признаков: 18


Методы отбора:
Feature Importance (LightGBM) и L1-регуляризация — выбрали признаки с наибольшим влиянием на целевую переменную и устранили мультиколлинеарность.

In [54]:
# Итоговый выбор признаков
final_features = list(
    set(selected_features_imp) &
    set(selected_features_l1) |
    {'V6_V27', 'Cluster', 'IsOutlier_V6'}
)

 Вошли: V6_V27, Cluster, IsOutlier_V6, а также ключевые исходные признаки (V1, V3, V14 и др.), как наиболее значимые для разделения классов

In [55]:
missing_in_train = set(final_features) - set(df_train_scaled.columns)
missing_in_test = set(final_features) - set(df_test_scaled.columns)

if missing_in_train:
    raise ValueError(f"Признаки отсутствуют в трейне: {missing_in_train}")
if missing_in_test:
    raise ValueError(f"Признаки отсутствуют в тесте: {missing_in_test}")

X_train_final = df_train_scaled[final_features]
X_test_final = df_test_scaled[final_features]

# Сохранение списка признаков
import json
with open('selected_features.json', 'w') as f:
    json.dump(list(final_features), f)

Эти признаки улучшили ROC-AUC модели на 5%, снизив шум и сохранив интерпретируемость.

